core/commonMain/src/kotlinx/serialization/PolymorphicSerializer.kt

```diff
@@ -67,14 +67,28 @@ import kotlin.reflect.*
  */
 @OptIn(ExperimentalSerializationApi::class)
 public class PolymorphicSerializer<T : Any>(override val baseClass: KClass<T>) : AbstractPolymorphicSerializer<T>() {
-    public override val descriptor: SerialDescriptor =
+
+    @PublishedApi // should we allow user access to this constructor?
+    internal constructor(
+        baseClass: KClass<T>,
+        classAnnotations: Array<Annotation>
+    ) : this(baseClass) {
+        _descriptor.annotations = classAnnotations.asList()
```

Comment:
Wouldn't it be better to make this the primary constructor and have the current primary become secondary? That also means that the descriptor can remain immutable.

Reply:
Unfortunately, old (e.g. 1.5.10) plugins are written in the way that they always pick the primary constructor. Changing its signature would break them

Reply:
It's worth dropping a comment here, especially when we do not have such compatibility tests.

Also, it forces us to use this pretty crafty hack. Do we have a chance to change that in the future?

Reply:
I think this hack can be fixed next time we raise minimal Kotlin version